### PR TITLE
[Merged by Bors] - feat: algebraic hierarchy on DirectLimit

### PR DESCRIPTION
**Pull request by alreadydone**
**Order/DirectedInverseSystem.lean**: define the DirectLimit of a DirectedSystem of types as a quotient (by a setoid (equivalence relation)) of the Sigma type (disjoint union). Provide the `induction` lemmas and `map`/`lift` constructions used to define algebraic operations on DirectLimit.

New file **Algebra/Colimit/DirectLimit.lean**: the first 400 lines are boilerplate code that defines algebraic instances on DirectLImit from magma (Mul) to Field. To make everything "hom-polymorphic", the DirectedSystem consists of FunLikes rather than plain/unbundled functions, and we use algebraic hom typeclasses (e.g. LinearMapClass, RingHomClass) everywhere. The remaining 70 lines essentially shows that DirectLimit is the colimit in the categories of Module and Ring.

**Algebra/DirectLimit.lean**: renamed to **Algebra/Colimit/ModuleRing.lean**. This file was originally added @kckennylau 5 years ago in [mathlib3#754](https://github.com/leanprover-community/mathlib3/pull/754) and defines `Module.DirectLimit` and `Ring.DirectLimit` as quotients of the universal objects (DirectSum and FreeCommRing). This definition is more general and suitable for arbitrary colimits, but makes it very cumbersome to prove [`of.zero_exact`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/DirectLimit.html#Ring.DirectLimit.of.zero_exact) lemmas (130+ lines!); for DirectLimit constructed as quotients of the disjoint union, these are just by definition. We golf these proofs by constructing isomorphisms between `Module/Ring.DirectLimit` and `_root_.DirectLimit`. I have more work done that generalizes most lemmas there to work for arbitrary colimits, that's why I put it under a new Colimit directory.

**Data/Int/Cast/Lemmas.lean**: add `map_intCast'`
**Data/Nat/Cast/Basic.lean**: golf a lemma
**Order/Directed.lean**: add a lemma
**ModelTheory/DirectLimit.lean**: change two lemmas to aliases of the general FunLike lemmas

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Int/Cast/Lemmas.lean [hunk 1]
```diff
@@ -220,6 +220,11 @@ theorem eq_intCast' [AddGroupWithOne α] [FunLike F ℤ α] [AddMonoidHomClass F
     ∀ n : ℤ, f n = n :=
   DFunLike.ext_iff.1 <| (f : ℤ →+ α).eq_intCastAddHom h₁
 
+/-- This version is primed so that the `RingHomClass` versions aren't. -/
+theorem map_intCast' [AddGroupWithOne α] [AddGroupWithOne β] [FunLike F α β]
```
Comment (jcommelin) [inline Mathlib/Data/Int/Cast/Lemmas.lean:224]: Is this not a strict generalization?

File: Mathlib/Algebra/Colimit/ModuleRing.lean [hunk 2]
```diff
@@ -237,85 +227,36 @@ end functorial
...
-and otherwise it is the zero map. -/
-noncomputable def totalize (i j) : G i →ₗ[R] G j :=
-  if h : i ≤ j then f i j h else 0
+variable [Nonempty ι] [IsDirected ι (· ≤ ·)] [DecidableEq ι] [DirectedSystem G (f · · ·)]
+open _root_.DirectLimit
 
-variable {G f}
+/-- The direct limit constructed as a quotient of the direct sum is isomorphic to
+the direct limit constructed as a quotient of the disjoint union. -/
+def linearEquiv : DirectLimit G f ≃ₗ[R] _root_.DirectLimit G f :=
```
Comment (jcommelin) [inline Mathlib/Algebra/Colimit/ModuleRing.lean:237]: So now we have two definitions of `DirectLimit`, whereas we had only one in the past, right?
One is a bit more general, but the other has better defeqs and is easier to work with. Do I understand that correctly?

I think part of your PR commit message contains useful commentary that should also be embedded in a module docstring in one of these files.

Comment (alreadydone) [reply Mathlib/Algebra/Colimit/ModuleRing.lean:237]: There were [Module.DirectLimit](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/DirectLimit.html#Module.DirectLimit), [AddCommGroup.DirectLimit](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/DirectLimit.html#AddCommGroup.DirectLimit) and [Ring.DirectLimit](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/DirectLimit.html#Ring.DirectLimit), and I'm now adding DirectLimit in the root space. The three existing `DirectLimit`s are in fact constructions of general colimits in the respective categories, while mine is restricted to directed colimits but applies to all kinds of algebraic structures. The point is that when the preorder is directed, the underlying type of all DirectLimits can be taken to be the same, no matter what algebraic structure you consider; in fact I simply take the DirectLimit in Type* and put all kinds of algebraic structures on it. The equivalence relation on that defines the DirectLimit in Type* is easy to understand, and is used to golf the `of.zero_exact` lemmas.

I'll see what part of the PR message is appropriate to include in the module docstrings.

Comment (jcommelin) [reply Mathlib/Algebra/Colimit/ModuleRing.lean:237]: Thanks, this message is also very useful. So please also add a version of that info to the module docstring.

File: Mathlib/Data/Int/Cast/Lemmas.lean (hunk 1 above)
Comment (alreadydone) [reply Mathlib/Data/Int/Cast/Lemmas.lean:224]: Do you mean `map_intCast`? That one assumes RingHomClass so there's no need to supply a proof of `f 1 = 1`. This new lemma is in parallel to [map_natCast'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Cast/Basic.html#map_natCast').

File: Mathlib/Algebra/Colimit/ModuleRing.lean (hunk 2 above)
Comment (alreadydone) [reply Mathlib/Algebra/Colimit/ModuleRing.lean:237]: Please check [2122ea8](https://github.com/leanprover-community/mathlib4/pull/19893/commits/2122ea8df7114d975d6bd6562e2b09032afb75dc); is there anything I should add?

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**